src/server/uds.rs

```diff
@@ -0,0 +1,233 @@
+                    return Ok(Async::Ready(Some(AddrStream::new(socket, addr))));
+                },
+                Ok(Async::NotReady) => return Ok(Async::NotReady),
+                Err(e) => {
+                    // Connection errors can be ignored directly, continue by
+                    // accepting the next request.
+                    if is_connection_error(&e) {
+                        debug!("accepted connection already errored: {}", e);
+                        continue;
+                    }
+
+                    if self.sleep_on_errors {
+                        // Sleep 1s.
+                        let delay = Instant::now() + Duration::from_secs(1);
+                        let mut timeout = Delay::new(delay);
+
+                        match timeout.poll() {
+                            Ok(Async::Ready(())) => {
+                                // Wow, it's been a second already? Ok then...
+                                error!("accept error: {}", e);
+                                continue
+                            },
+                            Ok(Async::NotReady) => {
+                                error!("accept error: {}", e);
+                                self.timeout = Some(timeout);
+                                return Ok(Async::NotReady);
+                            },
+                            Err(timer_err) => {
+                                error!("couldn't sleep on error, timer error: {}", timer_err);
+                                return Err(e);
+                            }
+                        }
+                    } else {
+                        return Err(e);
+                    }
+                },
+            }
+        }
+    }
+}
+
+/// This function defines errors that are per-connection. Which basically
+/// means that if we get this error from `accept()` system call it means
+/// next connection might be ready to be accepted.
+///
+/// All other errors will incur a timeout before next `accept()` is performed.
+/// The timeout is useful to handle resource exhaustion errors like ENFILE
+/// and EMFILE. Otherwise, could enter into tight loop.
+fn is_connection_error(e: &io::Error) -> bool {
+    match e.kind() {
+        io::ErrorKind::ConnectionRefused |
+        io::ErrorKind::ConnectionAborted |
+        io::ErrorKind::ConnectionReset => true,
+        _ => false,
+    }
+}
+
+impl fmt::Debug for UnixAddrIncoming {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("UnixAddrIncoming")
+            .field("addr", &self.addr)
+            .field("sleep_on_errors", &self.sleep_on_errors)
+            .finish()
+    }
+}
+
+mod addr_stream {
+    use std::io::{self, Read, Write};
+    use std::os::unix::net::SocketAddr;
+    use bytes::{Buf, BufMut};
+    use futures::Poll;
+    use tokio_uds::UnixStream;
+    use tokio_io::{AsyncRead, AsyncWrite};
+
+
+    #[derive(Debug)]
+    pub struct AddrStream {
```

Comment:
The name `AddrStream` probably makes less sense for UDS, it could be `UnixStream` or `UdsStream` or something.